### ISSUE TITLE
Unify test and developer environments

### ISSUE THREAD

**Issue by aanm** [cilium/cilium#2900]
Requirements for development environment:
 - [ ] Be able to set up any kubernetes version "on the fly", i.e. without re-provision the VM from scratch.
     - It can be achieved by running `NWORKERS=1 RELOAD=1 K8S_VERSION=1.X ./contrib/vagrant/start.sh`
 - [x] Be able to contact both VMs from outside. the current `NFS=1` option creates an interface exposed to the host.
 - [x] Both development and test environments should use the same VM names #2826 
     - `cilium-k8s-master` -> `k8s1`
     - `cilium-k8s-node-2` -> `k8s2`
     - `cilium-master` -> `runtime1`
     - `cilium-node-2` -> `runtime2`
 - [ ] By default, run cilium in baremetal mode (not in a daemonset / containers) 
 - [ ] Make use of `direnv` to store the environment variables in the developer's directory.
     - This allows the developer to run `./contrib/vagrant/start.sh` and `vagrant ssh k8s1` without worrying on setting up `NWORKERS=1 RELOAD=1 K8S=1 NFS=1` for every command
 - [x] Use same vagrant box as test environment #2826 

Requirements for test environments:  
- [ ] Run any kubernetes version
- [ ] Have different VM names depending on the kubernetes version set up
    - This allows to run multiple kubernetes versions in parallel

**Comment by raybejjani** [comment]
Does this mean that both runtime1 and k8s1 are masters?

**Comment by aanm** [comment]
>Does this mean that both runtime1 and k8s1 are masters?

Yes, well no. There's no meaning of "master" in non-k8s environments (`runtimeX`)

**Comment by joestringer** [comment]
* [ ] Easily share a list of directories from the host with the dev VM
  * This tends to be the case with the cilium directory today, but I also find it incredibly convenient to share eg. Linux sources, or iproute2 sources.
* [ ] Ability to make on the host and not have to make inside the VM
  * I tend to build on the host where I have nice tooling set up, then only run `make reload` to install/restart cilium inside the VM. Running bazel inside the VM means keeping another cache up to date, and consumes significant hard disk in the VM image (which causes other problems...)
* [ ] A recent version of `llvm-objdump`. We should consider keep the `clang` / `llvm` versions fairly old so that we can test with dumber versions of the compiler, but the ability to decode & debug BPF ELFs in the VM would be nice to have.

**Comment by raybejjani** [comment]
> Yes, well no. There's no meaning of "master" in non-k8s environments (runtimeX)

Oh, I didn't realise those were non-k8s. Are they generic boxes then? or could they have a name like docker/mesos/my-favourite-thingX ? I realise we've been using runtime already, so I don't want to turn this into a use-this-name debate, I'm mostly curious (and I wonder where I should have looked to understand it)

**Comment by aanm** [comment]
Based on this discussion https://github.com/cilium/cilium/issues/2689 we are re-adding NFS support. it won't affect the current shared folders integration
